src/bridge/cockpitdbusjson.c

```diff
@@ -2780,26 +2802,34 @@ on_connection_ready (GObject *source,
                      GAsyncResult *result,
                      gpointer user_data)
 {
-  CockpitDBusJson *self = COCKPIT_DBUS_JSON (user_data);
+  gchar *key = user_data;
+  GDBusConnection *connection;
   GError *error = NULL;
+  GPtrArray *subscribers;
 
-  self->connection = g_dbus_connection_new_for_address_finish (result,
-                                                               &error);
-  process_connection (self, error);
-  g_object_unref (self);
-}
+  connection = g_dbus_connection_new_for_address_finish (result, &error);
 
-static void
-on_bus_ready (GObject *source,
-              GAsyncResult *result,
-              gpointer user_data)
-{
-  CockpitDBusJson *self = COCKPIT_DBUS_JSON (user_data);
-  GError *error = NULL;
+  /*
+   * Notify all CockpitDBusJson objects from the same group that were
+   * also waiting for this connection
+   */
+  subscribers = g_hash_table_lookup (pending_group_connections, key);
+  for (guint i = 0; i < subscribers->len; i++)
+    {
+      CockpitDBusJson *channel = g_ptr_array_index (subscribers, i);
 
-  self->connection = g_bus_get_finish (result, &error);
-  process_connection (self, error);
-  g_object_unref (self);
+      if (connection)
+        channel->connection = g_object_ref (connection);
+
+      process_connection (channel, error);
+    }
+
+  g_hash_table_remove (pending_group_connections, key);
+
+  if (connection)
+    g_hash_table_insert (group_connections, g_strdup (key), connection);
```

Comment:
This can't be done in finalize, as other instances might still use the connections. I've added a weak notify on the connection object instead.